### ISSUE TITLE
force-dynamic no longer opts out of Data Caching since v14.2.0

### ISSUE THREAD

**Issue by mirkonasato** [vercel/next.js#65170]
### Link to the code that reproduces this issue

https://github.com/mirkonasato/force-dynamic-issue

### To Reproduce

1. Build and start the app in production mode
2. Load the home page a few times and see if the data changes

### Current vs. Expected behavior

Up to and including `next@14.1.1`, reloading the page fetches new data. This is also what the documentation says on [Data Fetching, Caching, and Revalidating](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching):

> fetch requests are not cached if:
> - [...]
> - The const dynamic = 'force-dynamic' route segment option is used.

From `next@14.2.0` up to at least `next@14.2.3` this is no longer the case: the page is re-rendered but appears to reuse data from the fetch cache.

See the 30-seconds video attached for a demo.

https://github.com/vercel/next.js/assets/133967/3f68f87f-13e0-48a6-804c-00a2b1b310e5

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:49 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6020
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 20.12.2
  npm: 10.5.0
  Yarn: N/A
  pnpm: 9.0.6
Relevant Packages:
  next: 14.2.3 // Latest available version is detected (14.2.3).
  eslint-config-next: N/A
  react: 18.3.1
  react-dom: 18.3.1
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next start (local)

### Additional context

I can't find anything in the [v14.2.0 release](https://github.com/vercel/next.js/releases/tag/v14.2.0) notes explaining this change.

Haven't tried how it works on Vercel; if the app doesn't seem to work locally I wouldn't deploy it to Vercel anyway.

**Comment by samcx** [comment]
@mirkonasato I was not able to replicate this on the latest canary → https://github.com/vercel/next.js/releases/tag/v14.3.0-canary.31 — can you verify on the latest canary as well?

**Comment by mirkonasato** [comment]
I can confirm that with `v14.3.0-canary.31` things work again like they did before `v14.2.0`.

In fact, it seems like the issue was fixed in [v14.2.1-canary.4](https://github.com/vercel/next.js/releases/tag/v14.2.1-canary.4), most likely by #64145.

**Comment by samcx** [comment]
@mirkonasato That's great to hear—I will be closing this issue out then in the meantime. Hopefully it will come up in a stable release soon!
